drivers/serial/uart_mcux.c

```diff
@@ -40,6 +42,8 @@ static int uart_mcux_configure(const struct device *dev,
 	uint32_t clock_freq;
 	status_t retval;
 
+	pinctrl_set_pin_state(config->pincfg.states[0]);
```

Comment:
Using states by index is fragile:

- Indexes have a meaning now (`0` means `default` state in this case)
- This won't work even though DT representation is correct (assume this driver had PM support):
```
pinctrl-0 = <config for sleep>;
pinctrl-1 = <config for default>;
pinctrl-names = "sleep", "default";
```
- If the application needs to reconfigure pins, it also needs to know how driver expects them to be sorted.

Reply:
Sure, I wasn't suggesting not supporting by name, just that the index solution would be the lowest level.

Reply:
I think having 2 ways of setting states not being fully equivalent (index vs. id) will make runtime pinctrl support fragile. I'd personally not take this path for such little gain: number of states is low (1, 2, 3...) and they can use an integer based identifier, so lookup overhead is low.

Reply:
We can't require that pinctrl-names is set so we have to allow for cases in which only `pinctrl-N` is utilized.

Reply:
-1 on that, I think we need to make usage consistent, otherwise we'll end up with a mixed usage depending on what developer thinks it is more "efficient" without accounting for side-effects such as usability of the runtime case.

Reply:
> -1 on that, I think we need to make usage consistent, otherwise we'll end up with a mixed usage depending on what developer thinks it is more "efficient" without accounting for side-effects such as usability of the runtime case.

This goes back to the design point of "upstream" devicetree bindings.  As we start seeing more and more Cortex-A SoCs being supported, we are going to have more devicetree(s) that cross between Linux and Zephyr.  As such we can't impose rules like this that don't exist already.

Reply:
This is really about a design decision, not an imposed rule. I don't see how the states concept works without naming `pinctrl-N`. It's what you need to do in Linux, for example. If the API doesn't have a clear design and allows for these sort of "shortcuts" I don't think we can expect a consistent/portable behavior. Whether this is acceptable or not is another thing.